test/lib/pwa.js

```diff
@@ -157,31 +158,56 @@ describe('lib.pwa', () => {
     });
   });
 
+  describe('#fetchManifest', () => {
+    afterEach(() => {
+      simpleMock.restore();
+    });
+    it('Fetches manifest directly from MANIFEST_URL', () => {
+      // Mock libManifest
+      simpleMock.mock(libManifest, 'fetchManifest').resolveWith(manifest);
+      return libPwa.fetchManifest(pwa).should.be.fulfilled.then(fetchedManifest => {
+        assert.equal(fetchedManifest, manifest);
+        assert.equal(libManifest.fetchManifest.callCount, 1);
+      });
+    });
+    it('Fails directly and looks for manifest link on START_URL', () => {
+      // Mock libManifest
+      simpleMock.mock(libManifest, 'fetchManifest').rejectWith(new Error()).resolveWith(manifest);
+      simpleMock.mock(dataFetcher, 'fetchLinkRelManifestUrl').resolveWith(START_URL);
+      return libPwa.fetchManifest(new Pwa(START_URL, manifest))
+      .should.be.fulfilled.then(fetchedManifest => {
+        assert.equal(fetchedManifest, manifest);
+        assert.equal(libManifest.fetchManifest.callCount, 2);
+        assert.equal(dataFetcher.fetchLinkRelManifestUrl.callCount, 1);
+      });
+    });
+  });
+
   describe('#save (core logic)', () => {
     afterEach(() => {
       simpleMock.restore();
     });
     it('performs all the save steps', () => {
       // Mock findByManifestUrl and bd to avoid making real calls
```

Comment:
Your test setup requires a lot of mocks. This potentially makes your test very fragile. Too many mocks in a test is a sign that your method might be doing too much (and this why test are great - they can tell you a lot about your design). 

If I look at the name of the method you're testing, it's "save". However, your mocks tell me it's also updating the icon, updating lighthouse info and more. This a lot more than just save. Consider refactoring into a more pipeline based approach, e.g.:

```
const manifestProvider = new ManifestProvider(queryManifestFromDB, fetchManifest);
const manifestProcessors = [updatePwaIcon, updatePwaCache, updatePwaLighthouseInfo, savePwaToDb,...];

const manifestPipeline = new ManifestPipeline(manifestProvider, manifestProcessors);

manifestPipeline.execute(pwa);
```

ManifestPipeline.execute(...) could look something like this:

```
execute(pwa) {
  this._manifestProvider.get(pwa)
    .then(manifest =>  {
       pwa.manifest = manifest;
       return Promise.all(manifestProcessors.map(p => p.apply(pwa)));
    })
}
```

You can then test each ManifestProcessor in isolation and mock one or two ManifestProcessors for ManifestPipeline tests (without relying on any implementation details besides the ManifestProcessor contract). 

This has the positive side-effect that you no longer need to export methods for testing (which is your test trying to tell you that your method/class is doing too much).